### ISSUE TITLE
pssac cannot plot more than 64 SAC files

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/gmt#1549]
Command:
```
gmt pssac input.dat -JX10c/5c -R9/20/-2/2 -Baf > single.ps
```
Example SAC file and the input.dat is also attached: [input+sac.zip](https://github.com/GenericMappingTools/gmt/files/3597119/input%2Bsac.zip)

The input.dat has 64 lines, thus the command plot the same file 64 times. The command works, and I can convert the PS file to PDF format.

However, if I add one more SAC file (i.e. adding one more line to input.dat), the command generates a corrupted PS file. `gmt psconvert -A -P -Tf` reports some errors:
```
Error: /rangecheck in --setdash--
Operand stack:
   --nostringval--   0
Execution stack:
   %interp_exit   .runexec2   --nostringval--   setdash   --nostringval--   2   %stopped_push   --nostringval--   setdash   setdash   false   1   %stopped_push   1974   1   3   %oparray_pop   1973   1   3   %oparray_pop   1961   1   3   %oparray_pop   1817   1   3   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   setdash   --nostringval--   2   %stopped_push   --nostringval--   1818   2   4   %oparray_pop
Dictionary stack:
   --dict:726/1123(ro)(G)--   --dict:0/20(G)--   --dict:75/200(L)--   --dict:154/250(L)--
Current allocation mode is local
Last OS error: No such file or directory
Current file position is 305504
psconvert [ERROR]: System call [gs -q -dSAFER -dNOPAUSE -dBATCH -sDEVICE=bbox -DPSL_no_pagefill -dMaxBitmap=2147483647 -dUseFastColor=true 'single.ps' 2> './psconvert_58795c.bb'] returned error 256.
```
Any idea what happened?

**Comment by PaulWessel** [comment]
I cannot reproduce it.  Your input.dat in the zip had 65 lines and worked.  I added one more and still worked.

**Comment by seisman** [comment]
Yes, the input.dat attached has 65 lines, which gives me a corrupted PS file.

Attached is the PS file I got (with 65 lines) [single.ps.zip](https://github.com/GenericMappingTools/gmt/files/3598369/single.ps.zip).

I'm using 6.0.0_2d2645a_2019.09.09, and gs 9.28rc2. Can you check if you can convert the PS file to PDF?

**Comment by PaulWessel** [comment]
Hm, this rings a bell but not sure what it is.  I find that your PS has a line like this:

[0] 0 B

but it should be

[] 0 B

That line is written by PSL_setdash (PSL, NULL, 0);
If you look at PSL_setdash in postscriptlight.c I cannot understand how it can write [0] given NULL, but perhaps I am missing something.

**Comment by seisman** [comment]
Possibly written by `psl_putdash`?

**Comment by PaulWessel** [comment]
Yes, PSL_setdash calls psl_putdash I think.

**Comment by PaulWessel** [comment]
Hm, maybe we need to set text[0] = '\0' in psl_putdash at the start to make sure it is empty? Or try a debug statement that checks if pattern != NULL and strcmp (pattern, "0") is true?

**Comment by seisman** [comment]
Now it works for me but I did nothing, I even didn't rebuild my GMT. Maybe some variables are not correctly initialized.

**Comment by PaulWessel** [comment]
Yes, I have seen this before but not able to track it down.  It has to be an uninitialized variable. You are doing this on macOS right?

**Comment by seisman** [comment]
Yes, on macOS. I see `static char text[PSL_BUFSIZ]` in `psl_putdash`. Is it intended to be a `static` variable?

**Comment by PaulWessel** [comment]
I think this is a quite old decision and since setdash is not called so often that there are any time-savings here, perhaps you could try an experiment:

RUn the current sac command 100 times in a loop and see if it will recreate the crash
Assuming it does, replace the 
	static char text[PSL_BUFSIZ];
with
 char text[PSL_BUFSIZ] = {""};

rebuild and run that 100 times.  No more crashes?

**Comment by seisman** [comment]
Re-run the sac command 100 times with following scripts:
```
for i in $(seq 1 100); do
    echo $i
    gmt pssac input.dat -JX10c/5c -R9/20/-2/2 -Baf > pssac-${i}.ps
    gmt psconvert -Tf pssac-${i}.ps
done
```
It should generate 100 identical PS files, but actually not. This is an example diff:
```
$ diff pssac-5.ps pssac-80.ps
8c8
< %%CreationDate: Tue Sep 10 23:44:14 2019
---
> %%CreationDate: Tue Sep 10 23:44:28 2019
39414,39415d39413
< 0 W
< [0] 0 B
39469,39470d39466
< 4 W
< [] 0 B
```

Changing to `char text[PSL_BUFSIZ] = {""};` still gives crashes.

**Comment by PaulWessel** [comment]
OK, so it cannot be the (NULL, 0) arguments, it must be a general setpen argument that has uninitialized values or somehow can get corrupted.  I need to head home now but will look more tonight. WOuld you have time to try this:

In psl_putdash, add a check that if pattern != NULL and it equals "0", then print some message to stderr? Your look should then see a few of those.  If that is the case then I can maybe try to run this in Xcode and maybe it will stop at that if test once and I may get a clue as to why.

**Comment by seisman** [comment]
Do you mean `if (pattern != NULL && pattern[0] == '\0')`?

**Comment by seisman** [comment]
I added following statement before the `if (pattern && pattern[0]) {` test.
```
if (pattern != NULL && strcmp (pattern, "0")) fprintf(stderr, "##%s##\n", pattern);
```

Running the sac command gives variable outputs for different runs, e.g.
```
##A##
##J##
##S##
##\##
##e##
##n##
##w##
##�##
```
and 
```
##S##
##\##
##e##
##n##
##w##
##�##
```

**Comment by PaulWessel** [comment]
Hm, I got nothing.  Compiled in debug or release or ???

**Comment by seisman** [comment]
Both release and debug give me similar outputs.

**Comment by PaulWessel** [comment]
Interesting. Wonder what the difference is in our setups.  If I could reproduce the problem then I would run this in Xcode and place a stop and see what line in the module is responsible.

**Comment by joa-quim** [comment]
The bug is in pssac. On Win I fell in Dongdong's condition at first run. When n=64 the L[64] struct is full of rubish.

![Capture](https://user-images.githubusercontent.com/537321/64726587-e4e3da00-d4ce-11e9-9241-7d6134fbccd5.PNG)

**Comment by joa-quim** [comment]
Not good. Because ``Ctrl->In = 1``

![Capture](https://user-images.githubusercontent.com/537321/64727799-9421b080-d4d1-11e9-8f64-6ab3d6902d3f.PNG)

n_files becomes = 0
```
	/* Got a bunch of SAC files or one file in SAC format */
	if (n_files > 1 || (n_files == 1 && issac(files[0]))) {
		L = gmt_M_memory (GMT, NULL, n_files, struct SAC_LIST) ;
		for (n = 0; n < n_files; n++) {
			L[n].file = strdup (files[n]);
			L[n].position = false;
			L[n].custom_pen = false;
		}
	}
	else {    /* Must read a list file */
		size_t n_alloc = 0;
		char pen[GMT_LEN256] = {""}, file[PATH_MAX] = {""};
		struct GMT_RECORD *In = NULL;
		double x, y;
		gmt_set_meminc (GMT, GMT_SMALL_CHUNK);
		n_files = 0;
```

and nothing is allocated in #L566

```
			if (n_files == n_alloc) L = gmt_M_malloc (GMT, L, n_files, &n_alloc, struct SAC_LIST);
```

**Comment by seisman** [comment]
@joa-quim Nice catch. All L[n].custom_pen should be `false` in my case, but actually some are `true`, possibly due to incorrect initialization.

`n_files==n_alloc==0`, so `gmt_M_malloc` allocates `64*sizeof(struct SAC_LIST)` to L, and n_alloc is 64. I assume, all the 64 elements are initialized, thus L[n].custom_pen are all `false`.

When n_files=64, `n_files == n_alloc == 64`,  `gmt_M_malloc` allocates another `64*sizeof(struct SAC_LIST)` more memory to L, but it seems `gmt_M_malloc` doesn't initialize the newly-allocated memory.

**Comment by PaulWessel** [comment]
Yes, I have run into this a lot.  If you have structures with pointers then the initial malloc initializes all to zero, but later reallocs are tricky and you must explicitly initialize the new memory.  We do this several places in GMT, e.g. see line 1352 in x2sys.c:

`gmt_M_memset (&(B->head[old_n_alloc]), n_alloc - old_n_alloc, struct X2SYS_BIX_TRACK_INFO);	/* Set content of new space to NULL */
`

**Comment by joa-quim** [comment]
Not sure. As sc shows you allocated it with 0 size.

![Capture](https://user-images.githubusercontent.com/537321/64729356-e9ab8c80-d4d4-11e9-93a9-b31460a24dad.PNG)

**Comment by seisman** [comment]
I believe `n_alloc` means first-time allocation, then `gmt_M_malloc` allocate 64 memory.

**Comment by seisman** [comment]
> If you have structures with pointers then the initial malloc initializes all to zero, but later reallocs are tricky and you must explicitly initialize the new memory. 

I think you found the culprit.

**Comment by seisman** [comment]
@PaulWessel Is it possible to let `gmt_M_malloc` initialize the new memory? If not, it should be documented.

**Comment by joa-quim** [comment]
You should never allocate with n_alloc = 0.

**Comment by seisman** [comment]
But it says n_alloc is allowed:
https://github.com/GenericMappingTools/gmt/blob/11f521b495364f37bc646ce9fa6b6d5bd3713f54/src/gmt_memory.c#L599-L620

**Comment by PaulWessel** [comment]
It would be good to find a solution to this as we are bound to do the same thing again somewhere.
After 6.0.0 release.

**Comment by seisman** [comment]
Closed by #1554.
